xkcd_cli/iv.py

```diff
@@ -0,0 +1,407 @@
+import sys
+import termios
+import atexit
+from select import select
+from base64 import standard_b64encode
+from typing import Literal, Optional, Set, Tuple, Union
+
+Protocol = Union[
+    Literal["iterm"], Literal["kitty"], Literal["kitty+"], Literal["sixel"]
+]
+
+
+class IV:
+    protocols: Set[Protocol] = set(("iterm", "kitty", "kitty+", "sixel"))
+    protocol: Optional[Protocol]
+    kitty: Optional[bool] = None
+    ex_kitty: Optional[bool] = None
+    sixel: Optional[bool] = None
+    iterm: Optional[bool] = None
+
+    def __init__(self, protocol: Optional[Union[Protocol, Literal["auto"]]] = None):
+        self.libsixel = None
+        self.stdin_fd = sys.stdin.fileno()
+        self.saved_term = termios.tcgetattr(self.stdin_fd)
+        atexit.register(self.set_normal_term)
+        if protocol == "auto":
+            self.auto_protocol()
+        elif protocol in self.protocols:
+            self.protocol = protocol
+        else:
+            self.protocol = None
+
+    @staticmethod
+    def scale_fit(
+        w: int,
+        h: int,
+        ow: int,
+        oh: int,
+        aspect: bool = True,
+        up: bool = False,
+        debug: bool = False,
+    ):
+        """Utility function which calculates new dimension of image"""
+        if debug:
+            print(w, h, ow, oh, aspect)
+        a = ow / oh
+        if not up and w > ow:
+            w = ow
+        if not up and h > oh:
+            h = oh
+        if w > 0 and h > 0:
+            if aspect:
+                if w / h > a:
+                    return int(h * a), h
+                else:
+                    return w, int(w / a)
+            else:
+                return w, h
+        elif w > 0:
+            if aspect:
+                return w, int(w / a)
+            else:
+                return w, oh
+        elif h > 0:
+            if aspect:
+                return int(h * a), h
+            else:
+                return ow, h
+        else:
+            return ow, oh
+
+    # Send and escape sequence and read reply.
+    def set_normal_term(self) -> None:
+        termios.tcsetattr(self.stdin_fd, termios.TCSAFLUSH, self.saved_term)
+
+    def terminal_request(self, cmd: str, end: str) -> str:
+        new_term = termios.tcgetattr(self.stdin_fd)
+        new_term[3] = new_term[3] & ~termios.ICANON & ~termios.ECHO
+        termios.tcsetattr(self.stdin_fd, termios.TCSAFLUSH, new_term)
+        ret = ""
+        print(cmd, end="", flush=True)
+        dr, _, _ = select([sys.stdin], [], [], 0.2)
+        if dr != []:
+            while True:
+                c = sys.stdin.read(1)
+                ret += c
+                if c in end:
+                    break
+        self.set_normal_term()
+        return ret
+
+    # Methods to send show image in various protocols.
+    def kitty_remove_placement(self, p: int = -1) -> None:
+        sys.stdout.buffer.write(f"\033_Ga=d,i=-1,d=z,z={p},q=2\033\\".encode("ascii"))
+        sys.stdout.flush()
```

Comment:
@MatanZ This method is not in use any longer. Can we safely remove it?

Reply:
It was never used here, and there is no problem with removing it.. See my explanation above.

It may be used in the future if you want to make the program interactive (display comic and then the user presses 'n' to display the next comic). This function allows you to remove the image without scrolling it out (and maybe to display it later without sending it to the terminal again).